modules/ui/src/main/scala/lucuma/ui/forms/FormInputEV.scala

```diff
@@ -77,36 +81,73 @@ object FormInputEV {
-        val newValue = props.valGet
-        // Force new value from props if the prop changes (or we are initializing).
-        stateOpt match {
-          case Some(state) if newValue === state.prevValue => state
-          case _                                           => State(newValue, newValue)
+  class Backend[EV[_], A]($ : BackendScope[Props[EV, A], State]) {
+    private val outerRef = Ref[html.Element]
+
+    def getInputElement(e: html.Element): html.Input =
+      e.firstChild
+        .asInstanceOf[html.Element]
+        .childNodes(1)
+        .asInstanceOf[html.Element]
+        .firstChild
+        .asInstanceOf[html.Input]
+
+    def getCursor: CallbackTo[Option[(Int, Int)]] =
+      outerRef.get.map(getInputElement).map(i => (i.selectionStart, i.selectionEnd)).asCallback
+
+    def setCursor(cursor: (Int, Int)): Callback =
+      outerRef.get.map(getInputElement).map(i => i.setSelectionRange(cursor._1, cursor._2))
+
+    def setStateCursor(offset: Int): Callback =
+      getCursor
+        .map(oc => oc.map { case (start, end) => (start + offset, end + offset) })
+        .flatMap(oc => $.setStateL(State.cursor)(oc))
+
+    def clearStateCursor: Callback = $.setStateL(State.cursor)(None)
+
+    def setCursorFromState: Callback =
+      $.state.flatMap(s => s.cursor.map(setCursor).getOrElse(Callback.empty))
+
+    def onTextChange: ReactEventFromInput => Callback =
+      (e: ReactEventFromInput) => {
+        $.props.flatMap { props =>
+          // Capture the value outside setState, react reuses the events
+          val v = e.target.value
+
+          def setDisplayValue(s: String): Callback    = $.setStateL(State.displayValue)(s)
+          def setModelValue(oa:  Option[A]): Callback =
+            if (props.updateOn == UpdateValue.OnChange)
+              oa.fold(Callback.empty)(a =>
+                // Set both state.modelValue and props.value so we don't reset state when we don't want to
+                $.setStateL(State.modelValue)(props.format.reverseGet(a)) *>
+                  props.valSet(a) *> props.onChange(a)
+              )
+            else Callback.empty
+
+          val result = props.changeAuditor(v, props.format)
+          result match {
+            case AuditResult.Accept(oa)                  =>
+              clearStateCursor *> setDisplayValue(v) *> setModelValue(oa)
+            case AuditResult.NewString(newS, oa, offset) =>
+              setStateCursor(offset) *> setDisplayValue(newS) *> setModelValue(oa)
+            case AuditResult.Reject()                    => setStateCursor(-1)
+          }
         }
       }
-      .render { $ =>
-        val p = $.props
-        val s = $.state
+
+    def onBlur(c: ChangeCallback[A]): Callback = for {
+      s  <- $.state
+      v   = s.displayValue
+      p  <- $.props
+      cb <- p.format.getOption(v).map(a => p.valSet(a) *> c(a)).getOrEmpty
+    } yield cb
+
+    val OuterDiv = <.div()
```

Comment:
I couldn't figure out a way to get access to a `Ref` without adding an extra div.  Is there a way around this?

Reply:
There is a new scalajs-react with some changes about `Ref`
https://github.com/japgolly/scalajs-react/blob/master/doc/changelog/1.7.6.md#vdom-refs

I haven't look at the details but there maybe there something that helps.

Reply:
I'll take a look. Thanks.

Reply:
I was hopeful that the changes to untyped `Ref`s in the new scalajs-react would fix this, but it turns out I can't get a Ref to the FormInput, anyway, because it is a `Function` component.